### DISCUSSION TITLE
hardhat 系列文章

### DISCUSSION THREAD

**Discussion by yurenju** [lun-dao/LunDAO#87] [Ideas]
因為之前有了 @a2468834 發表的 [mainnet forking 文章](https://lundao.tech/blog/hardhat-forking-1)，想說我們可以來來規劃一份 hardhat 系列文章，希望透過一系列的文章來引導讀者從最基礎的認識 hardhat，到逐步可以建立起實際上可以在 production 環境使用的環境建制。

目前會先有個規劃，然後我自己會先從簡單的入門開始寫，至於比較深入的文章想說先規劃內容跟目標讀者，如果有興趣一起寫的可以幫忙認領撰寫 👍 

目前系列文章規劃的文章如下：
- hardhat 入門 (@yurenju 認領)
	- 內容：講解 hardhat 的目的、提供的基礎功能以及如何建立起一個基礎專案
	- 目標讀者：大略了解 solidity 開發，但對 hardhat 使用不熟悉
- 如何使用 mainnet forking 進行 production 測試 -> 已有 [mainnet forking 文章](https://lundao.tech/blog/hardhat-forking-1)
- （選用）如何撰寫 hardhat task
	- 內容：講解要怎麼撰寫 hardhat task 以及怎麼設定各種參數
	- 目標讀者：想要幫自己的 hardhat 環境新增一些常用的指令
	- 為什麼是選用：如果入門文章還塞的下的話應該不需要這篇
- 如何撰寫測試、設定 coverage 以及透過 Github Action 設置 CI
	- 內容：講解跟測試與 CI 相關的工作
	- 目標讀者：想要撰寫測試的開發者，但不知道 hardhat/waffle 有提供哪些方便的功能
- 如何透過 hardhat-deploy 進行 upgradable contract 佈署
	- 內容：講解 hardhat-deploy 並且講解真的 production 的專案如果需要 upgradable contract 要怎麼設置
	- 目標讀者：已經在開發智能合約，並且需要在 production 環境操作的開發者

大家可以看看針對 hardhat 有沒有什麼內容是特別想了解的，也可以加入到系列文章裡面！

**Comment by a2468834** [comment]
這個 hardhat-storage-layout 感覺滿好用的 🤔
https://github.com/aurora-is-near/hardhat-storage-layout

雖然說使用 solc.js 也能做到 (在 `contracts[contract.sol][contract].storageLayout` 裡面)
但是 hardhat-storage-layout 印出的結果有經過排版，比較好閱讀

不過遇到 proxy contract 時好像會有問題
之前 wiasliaw 有開 issue，但是暫時還沒有人回覆 😅

**Reply by wiasliaw** [reply]
這東西沒人維護了，先不用看

**Reply by a2468834** [reply]
留個記錄
雖然套件沒維護了，不過還是可以提一下關於 storage layout 的分析
```javascript
module.exports = {
    solidity: {
        compilers: [
            {
                version: "0.5.5",
                outputSelection: {
                    '*': {
                        '*': ['*']
                    }
                }
            },
        ]
    },
};
```

**Comment by wiasliaw** [comment]
提案：https://github.com/zemse/hardhat-tracer

**Reply by yurenju** [reply]
這個好像很實用！

**Comment by gasolin** [comment]
提議：hardhat-vyper https://hardhat.org/plugins/nomiclabs-hardhat-vyper.html

**Comment by a2468834** [comment]
提議：hardhat-etherscan https://hardhat.org/plugins/nomiclabs-hardhat-etherscan.html
一鍵上傳合約原始檔到 Etherscan，超方便

**Reply by a2468834** [reply]
留個記錄
```javascript
module.exports = {
    etherscan: {
        apiKey: process.env.Etherscan
    }
};
```

**Comment by yurenju** [comment]
避免忘記， @zjiekai  有提到一個工具 foundry 看起來會跟 hardhat 功能有部分重疊，感覺也很值得寫一篇文章 https://github.com/foundry-rs/foundry

**Reply by zjiekai** [reply]
我只是看到 [optimistic-specs](https://github.com/ethereum-optimism/optimistic-specs/tree/main/packages/contracts) 里在用。我其实完全不知道这是什么用的。

**Comment by wiasliaw** [comment]
<img width="433" alt="image" src="https://user-images.githubusercontent.com/37038019/166156583-15a219d3-918d-40be-bfa6-96d171406673.png">

[ref link](https://v2-docs.zksync.io/dev/guide/hello-world.html#initializing-the-project-deploying-smart-contract)
